### PR TITLE
Resolve EXC_BAD_ACCESS in GULNetworkURLSession via O(1) passive memory lifecycle cleanup

### PR DESCRIPTION
**Pull request by agustindeleon7**
### Context & Problem
The previous implementation for cleaning up released sessions in `sessionIDToFetcherMap` relied on a manual O(N) loop. In a highly concurrent environment, this approach introduced severe race conditions. Specifically, iterating and reading the weak property (`holder.session`) while the session object was in the middle of its destruction phase led to an `EXC_BAD_ACCESS` crash within `objc_loadWeakRetained`. 

### Solution
This PR completely removes the active O(N) iteration loop and replaces it with an O(1) passive, lifecycle-driven cleanup strategy using Objective-C Associated Objects.

**Key Architectural Changes:**
1. **Introduced `GULSessionDeallocTracker`**: We attach a lightweight tracker to the session using `objc_setAssociatedObject`. The cleanup logic is now intrinsically tied to the session's natural ARC lifecycle.
2. **Eliminated `objc_loadWeakRetained` Crashes**: Reading a weak reference of an object currently undergoing deallocation is unsafe. To mathematically guarantee this crash won't happen, the tracker holds a strong reference to its `GULNetworkURLSessionWeakHolder` wrapper. During dealloc, it uses direct pointer comparison (`currentDictionaryHolder == _holder`) to safely verify if the entry should be removed, bypassing weak reference evaluation completely.
3. **Deadlock Prevention**: When replacing an existing session, the old tracker is explicitly extracted and released outside of the `NSLock` scope. This prevents the old tracker's `-dealloc` from attempting to re-acquire the global lock, which would cause an unrecoverable deadlock.

**Latest Refinements & CI Stability:**
- **Static Analysis Fix**: Addressed a static analyzer "Dead Store" warning by explicitly casting `(void)oldTrackerToReleaseOutsideLock;` after the lock is released.
- **Maintainability**: Added robust inline comments explaining the deadlock prevention strategy to prevent future refactoring regressions, alongside an `NSParameterAssert(sessionID)` for clear nullability contracts.
- **Architectural Test Cleanup**: Removed `testSessionMapStress` from `GULMutableDictionaryTest`. This test was improperly architected (it stressed `GULNetworkURLSession` rather than the dictionary itself), which caused heavy artificial lock contention and consistent CI timeouts on simulated runner environments.
- **Deterministic Validation**: Added `testSessionPassiveRemovalOnDeallocation` to `GULNetworkTest`. This focused unit test leverages an `@autoreleasepool` block to explicitly and deterministically validate the ARC-based cleanup mechanism without introducing flaky thread starvation.

### Impact
* **Stability:** Eliminates the `EXC_BAD_ACCESS` crash.
* **Performance:** Reduces dictionary cleanup time from O(N) to O(1).
* **Reliability:** CI tests are now deterministic, fully compliant with Apple's `os_unfair_lock`/`NSLock` concurrency limits under test environments, and natively validate the new memory lifecycle.

### CODE REVIEW

**Review by KarinaFernandez90 — APPROVED** [review]

**Comment by paulb777** [comment]
Please address CI failures

**Comment by agustindeleon7** [comment]
> Please address CI failures

Hi @paulb777! Thanks for the reply. I will fix the CI failures.

**Comment by agustindeleon7** [comment]
@paulb777 I've fixed the static analysis issues in this PR. Apologies for the oversight!

Quick question: [This failing job](https://github.com/google/GoogleUtilities/actions/runs/28042603354/job/83015086229?pr=233) couldn't find a file because it was recently removed in firebase/firebase-ios-sdk#16195. Is it safe to ignore this failure?

**Comment by ncooke3** [comment]
> @paulb777 I've fixed the static analysis issues in this PR. Apologies for the oversight!
> 
> Quick question: [This failing job](https://github.com/google/GoogleUtilities/actions/runs/28042603354/job/83015086229?pr=233) couldn't find a file because it was recently removed in [firebase/firebase-ios-sdk#16195](https://github.com/firebase/firebase-ios-sdk/pull/16195). Is it safe to ignore this failure?

Good catch! That's safe to ignore that one.

**Comment by agustindeleon7** [comment]
> > @paulb777 I've fixed the static analysis issues in this PR. Apologies for the oversight!
> > Quick question: [This failing job](https://github.com/google/GoogleUtilities/actions/runs/28042603354/job/83015086229?pr=233) couldn't find a file because it was recently removed in [firebase/firebase-ios-sdk#16195](https://github.com/firebase/firebase-ios-sdk/pull/16195). Is it safe to ignore this failure?
> 
> Good catch! That's safe to ignore that one.

Thanks @ncooke3! 

I'm seeing another issue regarding the Xcode version in [this job run](https://github.com/google/GoogleUtilities/actions/runs/28048145104/job/83033844019?pr=233). Should the Xcode version in the workflow be updated?

**Comment by paulb777** [comment]
#234 for the test coverage CI issue

Looks like there are also CI issues for finding Xcode 16.2 on main

**Comment by paulb777** [comment]
Gemini review:

Here is a review of the pull request **"Resolve EXC\_BAD\_ACCESS in GULNetworkURLSession via O(1) passive memory lifecycle cleanup" (\#233)** based on the provided diff.

Overall, this is a very elegant and well-thought-out Objective-C solution. It correctly replaces an inefficient O(N) active polling mechanism with an O(1) passive event-driven approach.

Here is a detailed breakdown of the changes, the good parts, and a few suggestions for the author.

### **🌟 The Good**

* **Smart use of Associated Objects:** Using objc\_setAssociatedObject to tie the GULSessionDeallocTracker to the session object's lifecycle is a classic, highly effective Objective-C pattern. When the session deallocates, the tracker automatically deallocates with it, instantly triggering the dictionary cleanup.  
* **Performance Improvement:** Removing the enumerateKeysAndObjectsUsingBlock: iteration inside setSessionInFetcherMap:forSessionID: guarantees an O(1) insertion time, rather than an O(N) sweep every time a session is added.  
* **Deadlock Prevention:** The inclusion of oldTrackerToReleaseOutsideLock is the most critical and impressive part of this PR. The author correctly identified that if objc\_setAssociatedObject(..., nil, ...) triggers the dealloc of the old tracker *while the lock is still held*, the tracker's dealloc method will attempt to re-acquire the non-recursive NSLock, causing an immediate deadlock. Capturing the old tracker and letting it fall out of scope outside the lock perfectly mitigates this.  
* **Memory Graph is Safe:** There are no retain cycles introduced here. The hierarchy is: session (retains) ➡️ tracker (retains) ➡️ holder (weakly references) ➡️ session.

### **💡 Suggestions & Feedback for the PR**

While the logic is sound, the intent behind some of the concurrency handling is subtle and could easily be accidentally removed by a future maintainer who doesn't understand it.

I would recommend adding a comment to the PR asking the author to clarify the oldTrackerToReleaseOutsideLock logic.

**1\. Add an inline comment for deadlock prevention**

The variable oldTrackerToReleaseOutsideLock and the (void) cast at the end of the function are clever but cryptic. A future developer might see (void)oldTrackerToReleaseOutsideLock;, assume it's useless dead code, delete it, and inadvertently reintroduce a deadlock.

*Suggested comment for the PR:*

"Great catch on preventing the deadlock with oldTrackerToReleaseOutsideLock. Because this is a very subtle concurrency detail, could we add a brief comment explaining *why* this is necessary so it doesn't get accidentally refactored out in the future?

Something like:

// Retain the old tracker until the lock is released. If it deallocates inside the lock, its \-dealloc method will attempt to re-acquire the lock, resulting in a deadlock."

**2\. Nullability of sessionID**

In GULSessionDeallocTracker's \-initWithSessionID:holder:, the sessionID property is blindly copied: \_sessionID \= \[sessionID copy\];.

While setSessionInFetcherMap:forSessionID: checks if sessionID is nil early on (if (\!sessionID) { return; }), it might be worth adding an NSParameterAssert(sessionID) in the tracker's init method just to be absolutely safe and document the non-null expectation.

### **Verdict**

**Approve, with minor nitpicks.** The core architectural shift is solid, performant, and correctly handles concurrency edge cases. It just needs a tiny bit of documentation to protect its cleverness from future regressions.

**Comment by agustindeleon7** [comment]
> #234 for the test coverage CI issue
> 
> Looks like there are also CI issues for finding Xcode 16.2 on main

I checked the Firebase repo and saw that their workflow uses Xcode 26.4 ([reference](https://github.com/firebase/firebase-ios-sdk/blob/669d38c472ca86b01be761d8eb5f6e0717c921e7/.github/workflows/_build.yml#L29-L30)). 

Would it make sense to align our configuration to use this same version?

**Comment by ncooke3** [comment]
> > #234 for the test coverage CI issue
> > Looks like there are also CI issues for finding Xcode 16.2 on main
> 
> I checked the Firebase repo and saw that their workflow uses Xcode 26.4 ([reference](https://github.com/firebase/firebase-ios-sdk/blob/669d38c472ca86b01be761d8eb5f6e0717c921e7/.github/workflows/_build.yml#L29-L30)).
> 
> Would it make sense to align our configuration to use this same version?

Thanks, but no worries. We can ignore those failures for this PR. I'll re-trigger CI.

**Comment by paulb777** [comment]
@morganchen12 Would you review this PR, the associated issue https://github.com/firebase/firebase-ios-sdk/issues/16299, and the https://github.com/google/GoogleUtilities/pull/230 which may have led to this potential regression.

File: GoogleUtilities/Network/GULNetworkURLSession.m [hunk 1]
```diff
@@ -693,27 +734,29 @@ + (void)setSessionInFetcherMap:(GULNetworkURLSession *)session forSessionID:(NSS
...
+    objc_setAssociatedObject(session, kGULSessionTrackerKey, tracker,
+                             OBJC_ASSOCIATION_RETAIN_NONATOMIC);
   } else {
     [[[self class] sessionIDToFetcherMap] removeObjectForKey:sessionID];
   }
   [[self sessionIDToFetcherMapReadWriteLock] unlock];
+
+  // Retain the old tracker until the lock is released. If it deallocates inside the lock,
+  // its -dealloc method will attempt to re-acquire the lock, resulting in a deadlock.
+  (void)oldTrackerToReleaseOutsideLock;
```
Comment (morganchen12) [inline GoogleUtilities/Network/GULNetworkURLSession.m:759]: My understanding is this can be optimized out by the compiler; you could explicitly assign this variable to `nil` instead.

**Review by morganchen12 — APPROVED** [review]
LGTM, just had one improvement.

File: GoogleUtilities/Network/GULNetworkURLSession.m (hunk 1 above)
Comment (paulb777) [reply GoogleUtilities/Network/GULNetworkURLSession.m:759]: Not sure about that. The original code was a `nil` but it causes a warning. The compiler can't change behavior with optimization. I think the comment should be sufficient to keep this as is.

**Comment by paulb777** [comment]
Thanks for the solid contribution @agustindeleon7 !! Deferring to @ncooke3 for the merge.

**Review by paulb777 — APPROVED** [review]

File: GoogleUtilities/Network/GULNetworkURLSession.m (hunk 1 above)
Comment (agustindeleon7) [reply GoogleUtilities/Network/GULNetworkURLSession.m:759]: Hi @morganchen12 @paulb777 

I initially tried assigning it to `nil`, but it causes the static analyzer to throw a "Dead Store" warning since the value is never read after the assignment. 

This warning makes the CI fail when running `pod_lib_lint.rb` with the `--analyze` flag for tvOS:
`scripts/third_party/travis/retry.sh scripts/pod_lib_lint.rb GoogleUtilities.podspec --platforms=tvos --analyze`

Casting it to `(void)` ([see commit here](https://github.com/google/GoogleUtilities/pull/233/changes/4b2a1c880fe6f238e5c730ac614b8f9f93e4dd91)) is just a workaround to tell the analyzer that the unused value is intentional, silencing the warning while safely releasing the tracker outside the lock.

**Comment by ncooke3** [comment]
Thanks, @agustindeleon7! Running CI for one final time...

**Review by KarinaFernandez90 — APPROVED** [review]

**Comment by Mountain-Tao** [comment]
When will the new version be released?

**Comment by ncooke3** [comment]
> When will the new version be released?

@Mountain-Tao, this is now released as 8.1.2. 

cc: @agustindeleon7

**Comment by agustindeleon7** [comment]
> > When will the new version be released?
> 
> @Mountain-Tao, this is now released as 8.1.2.
> 
> cc: @agustindeleon7

Great! Thanks!

bolt
